tests/dagger/stores/testblockexc.nim

```diff
@@ -118,58 +134,63 @@ suite "BlockExc engine - 2 nodes":
       sendDontHave: false)
 
     peerCtx1.peerWants.add(entry)
-    check blockexc2.taskQueue.pushOrUpdateNoWait(peerCtx1).isOk
+    check blockexc2.engine.taskQueue.pushOrUpdateNoWait(peerCtx1).isOk
     await sleepAsync(100.millis)
 
     check blockexc1.engine.localStore.hasBlock(blk.cid)
 
   test "should get blocks from remote":
-    let blocks = await blockexc1.getBlocks(blocks2.mapIt( it.cid ))
-    check blocks == blocks2
+    let blocks = await allFinished(
+      blocks2.mapIt( blockexc1.getBlock(it.cid) ))
+    check blocks.mapIt( !it.read ) == blocks2
 
   test "remote should send blocks when available":
-    let blk = !bt.Block.new("Block 1".toBytes)
+    let blk = bt.Block.new("Block 1".toBytes)
 
     # should fail retrieving block from remote
-    check not await blockexc1.getBlocks(@[blk.cid])
+    check not await blockexc1.getBlock(blk.cid)
       .withTimeout(100.millis) # should expire
 
-    proc onBlocks(evt: BlockStoreChangeEvt) =
-      check evt.cids == @[blk.cid]
-      done.complete()
-
-    blockexc1.engine.localStore.addChangeHandler(onBlocks, ChangeType.Added)
-
     # first put the required block in the local store
-    blockexc2.engine.localStore.putBlocks(@[blk])
+    await blockexc2.engine.localStore.putBlock(blk)
 
     # second trigger blockexc to resolve any pending requests
     # for the block
-    blockexc2.putBlocks(@[blk])
+    await blockexc2.putBlock(blk)
```

Comment:
I may be able to answer my own question with further study of the code, and/or maybe it's something we can discuss later today, but what is the relationship/difference between:
```nim
await blockexc2.engine.localStore.putBlock(blk)
```
and
```nim
await blockexc2.putBlock(blk)
```

Reply:
They both hold a reference to the same `localStore`, this is so that we can alway check if the block is available locally when requested from the node (for example the rest endpoint) or "put" the block to disk when one arrives from the network.

Blockexc is an implementation of the `BlockStore` interface and so is `localStore`.

Reply:
To expand on this a little bit, the `BlockStore` is the base "interface" used by different components to store and retrieve blocks either from local storage (i.e. disk) or the network. 

Architecturally, there are 2 main BlockStores the `NetworkStore` and the `LocalStore`. The `NetworkStore` is backed by the `BlockExchange` mechanism which as the name suggests implements the exchange of blocks across network peers.

The `LocalStore` abstracts aways storing blocks on disk.

Components then, talk to the `NetworkStore` which encapsulates a `LocalStore` to request or store blocks. The flow is more or less as follows:

Requesting blocks:

- Call `getBlock()` on the `NetworkStore` which will first check the `LocalStore`
  - if a block exist locally then return it
  - if no, dispatch it to the `BlockExchange` engine which will request it from some network peer

Storing blocks:

- Call `putBlock()` on the `NetworkStore`
  - `NetworkStore` calls `putBlock()` on `LocalStore` which will write it to local disk
  - Notify `BlockExchange` that a new block is available so that it can update interested peers with it's availability